src/BaseDAO/CLI.hs

```diff
@@ -73,36 +79,98 @@ data DaoContractInfo =
   }
 
 daoContractRegistry :: [DaoContractInfo] -> ContractRegistry
-daoContractRegistry contracts = ContractRegistry . Map.fromList $
-  contracts <&> \DaoContractInfo{..} ->
+daoContractRegistry contracts = ContractRegistry $ mconcat
+  [ Map.fromList $ contracts <&> \DaoContractInfo{..} ->
     dciName ?:: ContractInfo
       { ciContract = DAO.baseDaoContract dciConfig
       , ciIsDocumented = True
       , ciStorageParser = Just $ daoStorageParser dciExtraParser
       , ciStorageNotes = Nothing
       }
 
+  , one $ "MetadataCarrier" ?:: ContractInfo
+      { ciContract = metadataCarrierContract
+      , ciIsDocumented = False
+      , ciStorageParser = Nothing
+      , ciStorageNotes = Nothing
+      }
+  ]
+
+------------------------------------------------------------------------
+-- TZIP-16 metadata
+------------------------------------------------------------------------
+
+tzip16MetadataParser :: Opt.Parser (TZIP16.MetadataMap BigMap)
+tzip16MetadataParser = do
+  -- The primary scenario we support - reference to a contract
+  -- with actual metadata
+  contract <- contractHostParser
+  key <- metadataKeyParser
+  return $ TZIP16.metadataURI (TZIP16.tezosStorageUri contract key)
+  where
+    contractHostParser :: Opt.Parser TZIP16.ContractHost
+    contractHostParser = do
+      address <-
+        mkCLOptionParser Nothing (#name .! "metadata-host-address")
+        (#help .! "Address of the contract storing actual metadata")
+      mChainId <- Opt.optional . fmap fromString $
+        mkCLOptionParser Nothing (#name .! "metadata-host-chain")
+        (#help .! "Chain where the contract storing actual metadata resides")
+      return $
+        maybe TZIP16.contractHost TZIP16.foreignContractHost mChainId address
+
+    metadataKeyParser :: Opt.Parser MText
+    metadataKeyParser =
+      mkCLOptionParser Nothing (#name .! "metadata-key")
+        (#help .! "Key in metadata host contract where baseDAO's metadata is located")
+
 ------------------------------------------------------------------------
 -- Contract registry
 ------------------------------------------------------------------------
 
+data AllCmdLnArgs
+  = ContractRegistryCmd CmdLnArgs
+  | PrintMetadata
+
+allArgsParser :: DGitRevision -> ContractRegistry -> Opt.Parser AllCmdLnArgs
+allArgsParser gitRev registry = asum
+  [ ContractRegistryCmd <$>
+      argParser registry gitRev
+
+  , Opt.hsubparser $
+      mkCommandParser "print-metadata" (pure PrintMetadata)
+        "Print known part of TZIP-16 metadata."
+  ]
+
+mkCommandParser
+  :: String
+  -> Opt.Parser a
+  -> String
+  -> Opt.Mod Opt.CommandFields a
+mkCommandParser commandName parser desc =
+  Opt.command commandName $ Opt.info parser $ Opt.progDesc desc
+
 -- | Put this to @main@ to make the executable serve given contracts.
 serveContractRegistry :: String -> DGitRevision -> ContractRegistry -> Version -> IO ()
 serveContractRegistry execName gitRev contracts version =
   withUtf8 $ do
```

Comment:
While we are at it, can you change it to `wrapMain` please and remove `(die . displayException)`?